File: lib/src/main/scala/edu/illinois/osl/akka/gc/protocols/monotone/GC.scala
package edu.illinois.osl.akka.gc.protocols.monotone

import scala.collection.mutable

object GC {
    type Ref
    class Entry (
        val created: mutable.ArrayBuffer[Ref],
        val sentCount: mutable.HashMap[Token, Int],
        val recvCount: mutable.HashMap[Token, Int]
    )
    type History
    class Shadow {
        /** A shadow is marked if it is potentially unblocked in the current history */
        val isMarked: Boolean = false
        /** 
         * A list of unreleased refobs pointing to this actor.
         */
        val incoming: mutable.HashMap[Token,RefobStatus] = mutable.HashMap()
        /**
         * A list of unreleased refobs pointing from this actor.
         */
        val outgoing: mutable.HashMap[Ref, Shadow] = mutable.HashMap()
    }

    def onEntry(entry: Entry, history: History): Unit = {
        for (fact <- entry.created) {
            // In each of these cases, we need to look up the target actor's 
            // shadow and update it.
            // NB: The actor might not have a shadow yet!
            // ...and in future, the shadow may represent an actor in a remote system
            // If the token doesn't already exist, create the initial pending refob
            // If it does, do nothing


        }
        // Here we also need to update those shadows
        for ((token, n) <- entry.sentCount) {
            ???
            // If it's nonexistent, use the initial active refob
            // If it's pending, activate it first
            // If it's active, add the count

            // If it's active...
                // If the token doesn't already exist, create the initial active refob
                // If it does, activate it. It cannot already be active or pending.
            // If it's deactivated...
                // Hmm... We should only process this fact after we have adjusted the 
                // send count.
        }
        for ((token, m) <- entry.recvCount) {
            //val shadow: Shadow = ??? // Look up this actor's shadow
            //val incoming = shadow.incoming
            //val status = incoming.getOrElse(token, initialPendingRefob)
            //val newStatus = addToRecvCount(status, m)
            //if (refobIsReleased(status)) {
            //    incoming -= token
            //}
            //else {
            //    incoming(token) = newStatus
            //}
        }
        // Adjust this actor's outgoing refobs: add activated refobs, remove
        // released ones.
        ???
    }

    type RefobStatus = Integer

    def refobStatus(n: RefobStatus): String = {
        if ((n & 1) == 0) // Status bit is off
            return s"ACTIVE with ${n/2} pending messages"
        else if (n < 0) // Status bit is on and the int is negative
            return s"PENDING with ${(n+1)/2} received messages"
        else if (n > 1) // Status bit is on and the int is positive
            return s"FINISHING with ${(n-1)/2} pending messages"
        else            // n == 1, status bit is on and the int is zero
            return "RELEASED"
    }
}

File: lib/src/main/java/edu/illinois/osl/akka/gc/protocols/monotone/GC.java
package edu.illinois.osl.akka.gc.protocols.monotone;

public class GC {
    /** The size of each array in an entry */
    static int ARRAY_MAX = 16; // Need to use a power of 2 for the receive count

    public static void processEntry(Entry entry) {
        // Created refs
        for (int i = 0; i < ARRAY_MAX; i++) {
            if (entry.created[i] == null) break;
            if (entry.created[i].token().isDefined()) {
                Token token = entry.created[i].token().get();
                // Update the target actor's shadow.
                Shadow shadow = token.targetShadow();
                // If the token doesn't already exist, create the initial pending refob.
                // If it does, do nothing.
                shadow.incoming.putIfAbsent(token, RefobStatus.initialPendingRefob);
            }
            else {
                entry.shadow.isRoot = true;
                // TODO I'm assuming actors only create root refs to themselves
            }
        }
        // Sent info
        for (int i = 0; i < ARRAY_MAX; i++) {
            if (entry.sendTokens[i] == null) break;
            Token token = entry.sendTokens[i];
            short info = entry.sendInfos[i];

            boolean isActive = RefobInfo.isActive(info);
            boolean isDeactivated = !isActive;
            short sendCount = RefobInfo.count(info);

            Shadow targetShadow = token.targetShadow();
            int status = targetShadow.incoming.getOrDefault(token, RefobStatus.initialPendingRefob);
            boolean wasPending = RefobStatus.isPending(status);
            status = RefobStatus.addToSentCount(status, sendCount);
            if (isActive)
                status = RefobStatus.activate(status); // idempotent
            else
                status = RefobStatus.deactivate(status);
            boolean isReleased = RefobStatus.isReleased(status);

            // Adjust this actor's outgoing refobs: add activated refobs, remove released ones.
            if (isActive && wasPending) {
                // This entry must be the first one that mentions the refob
                Shadow prev = entry.shadow.outgoing.put(token, targetShadow);
                assert(prev == null);
            }
            else if (isDeactivated) {
                // This entry must be the one that deactivated the refob
                Shadow prev = entry.shadow.outgoing.remove(token);
                assert(prev == null);
            }
            if (isReleased) {
                targetShadow.incoming.remove(token);
            }
        }
        // Receive counts
        for (int i = 0; i < ARRAY_MAX; i++) {
            if (entry.recvTokens[i] == null) break;
            Token token = entry.recvTokens[i];
            short count = entry.recvCounts[i];
            Shadow shadow = entry.shadow;
            int status = shadow.incoming.getOrDefault(token, RefobStatus.initialPendingRefob);
            int newStatus = RefobStatus.addToRecvCount(status, count);
            if (RefobStatus.isReleased(newStatus)) {
                shadow.incoming.remove(token);
            }
            else {
                shadow.incoming.put(token, newStatus);
            }
        }
    }
}
